### ISSUE TITLE
Move invalid accumulation checks to `persist`

### ISSUE THREAD

**Issue by vmarcos** [MaterializeInc/materialize#24342]
### Feature request

Presently, we perform invalid accumulation detection in Compute in reduction operators. This is because reductions are operators that need to roll up a snapshot per key, and can then verify if the multiplicities that we end up with are negative (and thus invalid / violating a multi-set).

In a number of ways, these checks are best-effort. Only some reduction types are able to enforce checks and errors are only surfaced when we issue queries or maintain dataflows that include these operators. So invalid accumulations might still go undetected.

Additionally, the code of the reduce operators in rendering is made substantially more complex due to invalid accumulation detection. Also, future improvements to reduction performance may be impacted by the need to maintain these checks.

This issue is a request to move the invalid accumulation detection logic from Compute to `persist`. The request is based on the observation that any snapshot cannot include a negative accumulation. So whenever snapshots must be formed, e.g., due to compaction or when reading an initial snapshot via `persist_source`, the invalid accumulation checks could be introduced instead. These checks would then cast a wider net allowing us to catch invalid accumulations even when reductions are not present.

**Comment by benesch** [comment]
An idea for consideration: `persist` is arguably not the right place to enforce the "no negative accumulations" invariant, as `persist` is totally happy to deal in negative multiplicities. It seems like this is either a STORAGE (to the extent that STORAGE still exists as a layer on top of `persist`) or a COMPUTE invariant.

If it's a STORAGE invariant, then I think you can make the argument that `persist_source` is part of STORAGE and not of `persist`, and so it makes sense to enforce the invariant there.

If it's a COMPUTE invariant, then I think there's a world where compute could have a `no_negative_accumulations_enforcing_persist_source` that _wraps_ a `persist_source` and adds the invariant checking. This is maybe stressful to do today, as you'd need to force a consolidation where it might not have previously existed, but if `persist` learned to offer consolidate on read ... I think it falls out naturally?

Even in a world where this is a STORAGE invariant and not a COMPUTE invariant, possibly the right implementation is 1) support consolidate-on-read, 2) add a very straightforward check at the very end of `persist_source`.

cc @frankmcsherry @antiguru @danhhz

**Comment by danhhz** [comment]
Some initial brainstorming... I think the possibilities for where persist could do this check fall into two categories: on read or on write.

On read:
- Before we have consolidate on read, persist_source doesn't have any guarantees about equal keys being on the same worker, so the best we could do would be to check if the total sum of diffs of the snapshot was negative (which would still require an exchange).
- After we have consolidate on read, we could do this on a per key basis, yes.
- As far as I can tell, both of these seem just as easy/efficient to implement as operators that run after persist_source?

On write (which includes internal persist compaction):
- Persist could keep track of the total sum of diffs in each batch of the shard (very similar semantics to the batch in a dd Trace), this essentially gets us the same thing we could do in "On read" without consolidate_on_read.
- We could also check whether any keys in a "run" had a negative multiplicity. A persist batch is 1:N with runs, and the number of runs in a batch decrease as it goes through compactions, but we're not guaranteed that the batch with a lower of `[T::minimum()]` only has one run. So this would be best effort, but in a different way than the compute reductions best effort.
- It's not clear to me how persist would expose this information to the rest of mz and to the user.

**Comment by benesch** [comment]
> Edit: The thing I missed is that the issue description is talking about finding invalid retractions in the snapshot only!

Yes, sorry, exactly! My response was in the context of performing this check only for the snapshot. (I think @danhhz's was too.)

**Comment by benesch** [comment]
@frankmcsherry points out [on Slack](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1713787828937709?thread_ts=1713619148.645139&cid=C02PPB50ZHS):

> ...writing negative accumulations and advertising any unique key are inherently contradictory; the latter implies that all multiplicities are either 0 or 1

which, for me, points at STORAGE and COMPUTE being the right place to enforce/check this invariant, since both can be responsible for collections with unique keys—STORAGE for sources and tables, COMPUTE for indexes, and STORAGE/COMPUTE jointly for matviews.

**Comment by frankmcsherry** [comment]
For clarity, the ask is not about snapshot only tests; it asks to move all invalid accumulation checks to persist, and these include validating the collections at all times. When Marcos says "snapshot" he doesn't mean "the first time a dataflow is hydrated" he means "accumulate the updates up to some time", which `Reduce` does repeatedly for times, including those beyond the initial hydration timestamp.

> `persist` is arguably not the right place to enforce the "no negative accumulations" invariant, as `persist` is totally happy to deal in negative multiplicities.

Compute is also totally happy to deal in negative multiplicities. It's easy enough to clamp negative accumulations to zero. The pain exists is because historically compute also thinks that it is worth telling other folks if they've violated documented invariants, and no one else happens to check. Consequently, there's a bunch of parallel construction in compute because of the believed requirement that it provide errors on negative accumulations, but we could also just drop that belief and delete the code (and possibly have a worse product as a result, but one that is easier / more equitable to maintain locally, if harder to bisect problems).

**Comment by bkirwi** [comment]
There are a couple things that make invalid accumulation checks troublesome to do in Persist proper:
1. There are relatively few situations where we have an update in memory that we'd expect to be fully consolidated. (AFAICT, only when compacting or reading the first batch in the shard, and once Persist has compacted it down to a single run, which may take arbitrarily long.) This means any check at the Persist level will miss many real negative accumulations. "Consolidation on read" would solve this at read time, but is significantly more expensive than what we do today.
2. Persist doesn't know what a negative number is - it works for arbitrary semigroups / differential collections. We'd need to either specialize to i64 or plumb more traits through. (I don't hate the idea of specializing -- hard to imagine another diff type being useful at the Persist level -- but I know it's been controversial in the past.)
3. There are patterns where it would be useful to write down negative accumulations in a Persist shard, like some of the approaches for hiding errors in other shards. (Those involve writing a shard with negative accumulations and unioning it with another shard with positive accumulations before showing it to a user.)

I don't see any strong reason to avoid doing it in persist_source or persist_sink, which the cluster team owns, though! But that doesn't sound like what this issue is after.

**Comment by ggevay** [comment]
> There are patterns where it would be useful to write down negative accumulations in a Persist shard, like some of the approaches for hiding errors in other shards. (Those involve writing a shard with negative accumulations and unioning it with another shard with positive accumulations before showing it to a user.)

Could you say more about such patterns? Naively, I'd think that we could always just negate a collection when reading it (e.g., before unioning it with something else).

**Comment by bkirwi** [comment]
> Could you say more about such patterns?

Sure, though I should note that this is IMO my least compelling point in that list. 😅

Suppose I have a subsource that sometimes has data I don't like. (Perhaps it's has a bad row, or perhaps it's missing a row that it _should_ have.) We've talked about letting users write to subsources, though for reasons discussed in other issues that's a little bit fraught. An alternative would be to have a "diff" shard, where you store the updates that would take the "raw" subsource to "how it should be". This could include both positive accumulations (for rows to add) or negative ones (for rows to delete).

It is always possible to avoid negative accumulations at the Persist level... but then negating it and unioning it possibly reintroduces negative accumulations at a compute level, so nothing is gained!

**Comment by bkirwi** [comment]
One inexpensive thing we _could_ do is have `shard_source` tag parts with whether or not they're expected to be fully consolidated, and `persist_source` could assert that it never generates negative accumulations for those parts (modulo temporal filters etc.) But I don't think that's a high-coverage enough check to be willing to remove the assertions we make elsewhere.

**Comment by ggevay** [comment]
> Sure, though I should note that this is IMO my least compelling point in that list. 😅

Sorry, I was picking on this point, because this is the point that would affect whether we could disallow negative accumulations as an invariant (and not just where we should add checks for them), which I'm most interested in at the moment.

**Comment by benesch** [comment]
I also believe it's currently possible, in unsafe mode, to write:

```sql
CREATE TABLE t (a int);
INSERT INTO t SELECT 1 FROM repeat_row(-1);
```

which will produce a table with negative accumulations.
